packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -597,12 +609,14 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
         }
       } else {
         if (rightArrow && newSelection.extentOffset < _plainText.length) {
-          newSelection = newSelection.copyWith(extentOffset: newSelection.extentOffset + 1);
+          final int delta = _isLeadingSurrogate(text.codeUnitAt(newSelection.extentOffset)) ? 2 : 1;
```

Comment:
can you add a todo to indicate why the workaround and we should remove it when the character library is integrated?

Reply:
same for other place

Reply:
If i remember correctly, there are grapheme clusters that are more than two characters. for example 👨‍👩‍👦.  Are we going to handle them as well?

Reply:
> can you add a todo to indicate why the workaround

It's not a workaround. Preventing arrowing into, or deleting half of, a surrogate pair is correct handling of UTF-16.

> If i remember correctly, there are grapheme clusters that are more than two characters. for example 👨‍👩‍👦. Are we going to handle them as well?

Not in this PR; per the PR description:

> This makes the arrow key and delete key handling surrogate-aware
(although not grapheme-cluster-aware, which is a larger fix that is out
of scope here).

See also my answer to the comments on the PR.

But breaking other grapheme cluster doesn't (AFAIK) give you invalid UTF-16 that can render the text field permanently unusable.

Reply:
> It's not a workaround. Preventing arrowing into, or deleting half of, a surrogate pair is correct handling of UTF-16.
I disagree a little bit, isn't this going to be removed when this is refactored using the character library? Whether it is a workaround or not, we will want to refactor this part when the character library is ready. I think it will be good to add a todo to remind us that.
> 
> But breaking other grapheme cluster doesn't (AFAIK) give you invalid UTF-16 that can render the text field permanently unusable.

I see, that makes sense.

Reply:
> > It's not a workaround. Preventing arrowing into, or deleting half of, a surrogate pair is correct handling of UTF-16.
> 
> I disagree a little bit, isn't this going to be removed when this is refactored using the character library? Whether it is a workaround or not, we will want to refactor this part when the character library is ready. I think it will be good to add a todo to remind us that.

This PR has no impact on the fact that this code will change when the `characters` library is ready. Either with this change, or without this change, arrow key handling and delete handling currently don't handle arbitrary grapheme clusters correctly, and either way the code being touched here will be rewritten by a PR that uses `characters` to do grapheme handling (since that is the point of such a PR).

I'm not clear on how making code that's going to be rewritten either way less wrong until it's rewritten is a workaround, and I'm not seeing the benefit of adding a todo reminding us to do something that a) already needed to be done anyway, and b) and already has, IIUC, a design doc, a bug, and an in-progress PR.

Reply:
I see, I was over thinking it. The test will fail anyway if they forget to remove this part.